central/image/datastore/store/v2/postgres/store.go

```diff
@@ -0,0 +1,1027 @@
+package postgres
+
+import (
+	"context"
+	"time"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/mitchellh/hashstructure/v2"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stackrox/rox/central/image/datastore/store"
+	"github.com/stackrox/rox/central/image/datastore/store/common/v2"
+	"github.com/stackrox/rox/central/metrics"
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/logging"
+	ops "github.com/stackrox/rox/pkg/metrics"
+	"github.com/stackrox/rox/pkg/postgres"
+	"github.com/stackrox/rox/pkg/postgres/pgutils"
+	pkgSchema "github.com/stackrox/rox/pkg/postgres/schema"
+	"github.com/stackrox/rox/pkg/protocompat"
+	"github.com/stackrox/rox/pkg/search"
+	pgSearch "github.com/stackrox/rox/pkg/search/postgres"
+)
+
+const (
+	imagesTable                = pkgSchema.ImagesTableName
+	imageComponentsV2Table     = pkgSchema.ImageComponentV2TableName
+	imageComponentsV2CVEsTable = pkgSchema.ImageCvesV2TableName
+
+	getImageMetaStmt = "SELECT serialized FROM " + imagesTable + " WHERE Id = $1"
+
+	cursorBatchSize = 50
+)
+
+var (
+	log    = logging.LoggerForModule()
+	schema = pkgSchema.ImagesSchema
+)
+
+type imagePartsAsSlice struct {
+	image        *storage.Image
+	componentsV2 []*storage.ImageComponentV2
+	cvesV2       []*storage.ImageCVEV2
+}
+
+type timeFields struct {
+	createdAt            time.Time
+	firstImageOccurrence time.Time
+}
+
+// TODO(ROX-28222): Refactor logic operating on other tables out and up to the datastore layer.
+
+// New returns a new Store instance using the provided sql instance.
+func New(db postgres.DB, noUpdateTimestamps bool, keyFence concurrency.KeyFence) store.Store {
+	return &storeImpl{
+		db:                 db,
+		noUpdateTimestamps: noUpdateTimestamps,
+		keyFence:           keyFence,
+	}
+}
+
+type storeImpl struct {
+	db                 postgres.DB
+	noUpdateTimestamps bool
+	keyFence           concurrency.KeyFence
+}
+
+func (s *storeImpl) insertIntoImages(
+	ctx context.Context,
+	tx *postgres.Tx, parts *imagePartsAsSlice,
+	metadataUpdated, scanUpdated bool,
+	iTime time.Time,
+) error {
+	// First Image Occurrence and Created At are set based on the CVE itself, not the CVE
+	// within the image.  Since a CVE can occur multiple times within an image we can grab
+	// those times for the incoming data and set the times appropriately.  We will later go through the
+	// existing CVEs to make further adjustments if necessary to make sure we do not overwrite
+	// the times of previous occurrences.
+	cveTimeMap := make(map[string]*timeFields)
+	for _, cve := range parts.cvesV2 {
+		if val, ok := cveTimeMap[cve.GetCveBaseInfo().GetCve()]; ok {
+			if val.createdAt.After(cve.GetCveBaseInfo().GetCreatedAt().AsTime()) {
```

Comment:
If `createdAt` or `firstImageOccurrence` is nil for a CVE in some components but not in others, I think this logic would choose the nil timestamp as final since it translates to Jan 1970 or something. Is that okay? Or should we chose the min among non-nil timestamps?